// unpic/esm/src/transformers/cloudflareimages.js
import { toUrl } from "../utils.js";
const cloudflareImagesRegex = /https?:\/\/(?<host>[^\/]+)\/cdn-cgi\/imagedelivery\/(?<accountHash>[^\/]+)\/(?<imageId>[^\/]+)\/*(?<transformations>[^\/]+)*$/g;
const parseTransforms = (transformations) => Object.fromEntries(transformations?.split(",")?.map((t) => t.split("=")) ?? []);
const formatUrl = ({ host, accountHash, transformations = {}, imageId, }) => {
    const transformString = Object.entries(transformations).map(([key, value]) => `${key}=${value}`).join(",");
    const pathSegments = [
        host,
        "cdn-cgi",
        "imagedelivery",
        accountHash,
        imageId,
        transformString,
    ].join("/");
    return `https://${pathSegments}`;
};
export const parse = (imageUrl) => {
    const url = toUrl(imageUrl);
    const matches = [...url.toString().matchAll(cloudflareImagesRegex)];
    if (!matches.length) {
        throw new Error("Invalid Cloudflare Images URL");
    }
    const group = matches[0].groups || {};
    const { transformations: transformString, ...baseParams } = group;
    const { w, h, f, ...transformations } = parseTransforms(transformString);
    return {
        base: url.toString(),
        width: Number(w) || undefined,
        height: Number(h) || undefined,
        format: f,
        cdn: "cloudflare_images",
        params: { ...baseParams, transformations },
    };
};
export const generate = ({ base, width, height, format, params }) => {
    const parsed = parse(base.toString());
    const props = {
        transformations: {},
        ...parsed.params,
        ...params,
    };
    if (width) {
        props.transformations.w = width?.toString();
    }
    if (height) {
        props.transformations.h = height?.toString();
    }
    if (format) {
        props.transformations.f = format;
    }
    props.transformations.fit ||= "cover";
    return new URL(formatUrl(props));
};
export const transform = ({ url: originalUrl, width, height, format = "auto" }) => {
    const parsed = parse(originalUrl);
    if (!parsed) {
        throw new Error("Invalid Cloudflare Images URL");
    }
    const props = {
        ...parsed,
        width,
        height,
        format,
    };
    return generate(props);
};
