// unpic/script/src/transformers/cloudflareimages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = exports.generate = exports.parse = void 0;
const utils_js_1 = require("../utils.js");
const cloudflareImagesRegex = /https?:\/\/(?<host>[^\/]+)\/cdn-cgi\/imagedelivery\/(?<accountHash>[^\/]+)\/(?<imageId>[^\/]+)\/*(?<transformations>[^\/]+)*$/g;
const parseTransforms = (transformations) => Object.fromEntries(transformations?.split(",")?.map((t) => t.split("=")) ?? []);
const formatUrl = ({ host, accountHash, transformations = {}, imageId, }) => {
    const transformString = Object.entries(transformations).map(([key, value]) => `${key}=${value}`).join(",");
    const pathSegments = [
        host,
        "cdn-cgi",
        "imagedelivery",
        accountHash,
        imageId,
        transformString,
    ].join("/");
    return `https://${pathSegments}`;
};
const parse = (imageUrl) => {
    const url = (0, utils_js_1.toUrl)(imageUrl);
    const matches = [...url.toString().matchAll(cloudflareImagesRegex)];
    if (!matches.length) {
        throw new Error("Invalid Cloudflare Images URL");
    }
    const group = matches[0].groups || {};
    const { transformations: transformString, ...baseParams } = group;
    const { w, h, f, ...transformations } = parseTransforms(transformString);
    return {
        base: url.toString(),
        width: Number(w) || undefined,
        height: Number(h) || undefined,
        format: f,
        cdn: "cloudflare_images",
        params: { ...baseParams, transformations },
    };
};
exports.parse = parse;
const generate = ({ base, width, height, format, params }) => {
    const parsed = (0, exports.parse)(base.toString());
    const props = {
        transformations: {},
        ...parsed.params,
        ...params,
    };
    if (width) {
        props.transformations.w = width?.toString();
    }
    if (height) {
        props.transformations.h = height?.toString();
    }
    if (format) {
        props.transformations.f = format;
    }
    props.transformations.fit ||= "cover";
    return new URL(formatUrl(props));
};
exports.generate = generate;
const transform = ({ url: originalUrl, width, height, format = "auto" }) => {
    const parsed = (0, exports.parse)(originalUrl);
    if (!parsed) {
        throw new Error("Invalid Cloudflare Images URL");
    }
    const props = {
        ...parsed,
        width,
        height,
        format,
    };
    return (0, exports.generate)(props);
};
exports.transform = transform;
