// unpic/script/src/transformers/shopify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = exports.generate = exports.parse = void 0;
const utils_js_1 = require("../utils.js");
const shopifyRegex = /(.+?)(?:_(?:(pico|icon|thumb|small|compact|medium|large|grande|original|master)|(\d*)x(\d*)))?(?:_crop_([a-z]+))?(\.[a-zA-Z]+)(\.png|\.jpg|\.webp|\.avif)?$/;
const parse = (imageUrl) => {
    const url = (0, utils_js_1.toUrl)(imageUrl);
    const match = url.pathname.match(shopifyRegex);
    if (!match) {
        throw new Error("Invalid Shopify URL");
    }
    const [, path, size, width, height, crop, extension, format] = match;
    url.pathname = `${path}${extension}`;
    const widthString = width ? width : url.searchParams.get("width");
    const heightString = height ? height : url.searchParams.get("height");
    url.searchParams.delete("width");
    url.searchParams.delete("height");
    return {
        base: url.toString(),
        width: Number(widthString) || undefined,
        height: Number(heightString) || undefined,
        format: format ? format.slice(1) : undefined,
        params: { crop, size },
        cdn: "shopify",
    };
};
exports.parse = parse;
const generate = ({ base, width, height, format, params }) => {
    const url = (0, utils_js_1.toUrl)(base);
    (0, utils_js_1.setParamIfDefined)(url, "width", width, true, true);
    (0, utils_js_1.setParamIfDefined)(url, "height", height, true, true);
    (0, utils_js_1.setParamIfDefined)(url, "crop", params?.crop);
    (0, utils_js_1.setParamIfDefined)(url, "format", format);
    return url;
};
exports.generate = generate;
const transform = ({ url: originalUrl, width, height }) => {
    const parsed = (0, exports.parse)(originalUrl);
    if (!parsed) {
        return;
    }
    const props = {
        ...parsed,
        width,
        height,
    };
    return (0, exports.generate)(props);
};
exports.transform = transform;
