// unpic/script/src/transformers/directus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = exports.parse = void 0;
const utils_js_1 = require("../utils.js");
const parse = (imageUrl) => {
    const parsedUrl = (0, utils_js_1.toUrl)(imageUrl);
    const width = (0, utils_js_1.getNumericParam)(parsedUrl, "width");
    const height = (0, utils_js_1.getNumericParam)(parsedUrl, "height");
    const format = parsedUrl.searchParams.get("format") || undefined;
    const quality = (0, utils_js_1.getNumericParam)(parsedUrl, "quality") || undefined;
    let fit = parsedUrl.searchParams.get("fit") ||
        undefined;
    const withoutEnlargement = parsedUrl.searchParams.get("withoutEnlargement") === "true" || undefined;
    const transforms = parsedUrl.searchParams.get("transforms") || undefined;
    // if fit doesn't satisfy the type, it will be undefined
    if (fit && !["cover", "contain", "inside", "outside"].includes(fit)) {
        fit = undefined;
    }
    return {
        width,
        height,
        format,
        base: parsedUrl.toString(),
        params: {
            fit,
            quality,
            withoutEnlargement,
            transforms: transforms ? JSON.parse(transforms) : undefined,
        },
        cdn: "directus",
    };
};
exports.parse = parse;
const transform = ({ url: originalUrl, width, height, format }) => {
    const url = (0, utils_js_1.toUrl)(originalUrl);
    (0, utils_js_1.setParamIfDefined)(url, "width", width, true, true);
    (0, utils_js_1.setParamIfDefined)(url, "height", height, true, true);
    (0, utils_js_1.setParamIfDefined)(url, "format", format);
    (0, utils_js_1.setParamIfDefined)(url, "quality", (0, utils_js_1.getNumericParam)(url, "quality"), true);
    return url;
};
exports.transform = transform;
